File: miniapp/lib/nl.ts
// Natural-language member Q&A — TypeScript port of agent/src/nl.py's deterministic layer.
//
// HARD RULE (CLAUDE.md §1.3): this never moves money and never authorizes an action the
// contract wouldn't enforce. It only explains facts read from chain. Keep this file's
// logic in sync with agent/src/nl.py — that module remains the source of truth for the
// Python agent and its tests.

// NL handler model — small + fast, explanation only (docs/STACK.md). Never tool-enabled.
// DeepSeek (OpenAI-compatible) so the rephrasing runs on the key the deployment holds.
export const NL_MODEL = "deepseek-chat";

export const SYSTEM_PROMPT = `You are AjoAI's member assistant for a rotating savings circle \
(ajo/esusu/chama/stokvel) on Celo. You ONLY explain facts you are given about the circle. \
You NEVER promise to move money, change anyone's turn, waive a penalty, or take any action — \
the smart contract enforces all of that, not you. If asked to do something money-moving, \
explain what the contract rules are instead.

Answer in the SAME language the member used: English, Nigerian Pidgin, or Swahili. Keep it \
short (1-3 sentences), concrete, and money-accurate. Use the exact figures provided. Never \
invent numbers; if a fact is not in the context, say you don't have it.`;

/** Format a token amount with up to 2 decimals + symbol, e.g. "0.6 USDT" (not "0 units"). */
export function fmtUnits(amountWei: bigint, decimals: number, symbol = "units"): string {
  const base = 10n ** BigInt(decimals);
  const whole = amountWei / base;
  const frac = amountWei % base;
  let s = whole.toLocaleString("en-US");
  if (frac > 0n) {
    // two significant fractional digits
    const twoDp = (frac * 100n) / base;
    if (twoDp > 0n) s += "." + twoDp.toString().padStart(2, "0").replace(/0+$/, "");
  }
  return `${s} ${symbol}`;
}

/** Chain snapshot shape needed to derive a member's situation (mirrors CircleView). */
export type CircleSnapshot = {
  members: readonly string[];
  rotation: readonly string[];
  currentRound: number;
  recipient: string | null;
  recipientDelinquent: boolean;
  stateName: string;
  intendedPot: bigint;
  slots: number;
  symbol: string;
};

export type MemberFacts = {
  isMember: boolean;
  hasReceived: boolean;
  isDelinquent: boolean;
  yourRound: number | null;
  roundsUntilYourTurn: number | null;
  currentRecipient: string | null;
  state: string;
  intendedPotStr: string;
  joined: number;
  slots: number;
};

/** Pure: derive a member's situation from a chain snapshot (no LLM, no tx). */
export function factsFor(view: CircleSnapshot, member: string, tokenDecimals = 18): MemberFacts {
  const m = member.toLowerCase();
  const membersLc = view.members.map((a) => a.toLowerCase());
  const isMember = membersLc.includes(m);

  let yourRound: number | null = null;
  if (view.rotation.length) {
    const rotLc = view.rotation.map((a) => a.toLowerCase());
    const idx = rotLc.indexOf(m);
    if (idx !== -1) yourRound = idx;
  }

  const roundsUntilYourTurn = yourRound !== null ? Math.max(yourRound - view.currentRound, 0) : null;

  // hasReceived: not tracked per-member on the snapshot; infer for the common case.
  const hasReceived = yourRound !== null && view.currentRound > yourRound;

  return {
    isMember,
    hasReceived,
    isDelinquent: Boolean(view.recipient && view.recipient.toLowerCase() === m && view.recipientDelinquent),
    yourRound,
    roundsUntilYourTurn,
    currentRecipient: view.recipient,
    state: view.stateName,
    intendedPotStr: fmtUnits(view.intendedPot, tokenDecimals, view.symbol),
    joined: view.members.length,
    slots: view.slots,
  };
}

/** Deterministic, money-safe answer derived purely from chain facts. */
export function baselineAnswer(f: MemberFacts): string {
  // Non-member: answer about the circle's state instead of a flat rejection (the organizer who
  // has not joined yet, an invitee browsing, or someone viewing a finished circle all land here).
  if (!f.isMember) {
    if (f.state === "Forming") {
      return `This circle is still forming: ${f.joined} of ${f.slots} have joined. You are not in it yet. Tap Join to take a slot (you post a one-round security deposit, returned on clean completion).`;
    }
    if (f.state === "Completed") return "This circle has finished. Every member received their payout once and it is now complete.";
    if (f.state === "Defaulted") return "This circle has ended in default. Remaining funds were distributed to members who had not yet received.";
    if (f.state === "Dissolved") return "This circle was dissolved while still forming and all deposits were refunded.";
    return `This circle is active with ${f.slots} members; you are not a member of it.`;
  }
  // Terminal state (member): NEVER project a future round — the rotation is over. `currentRound` is
  // stale here, so the round arithmetic below would mis-tell a member "your payout is in N rounds"
  // for a circle that already ended (money-inaccurate, CLAUDE.md §8). State the outcome instead.
  if (f.state === "Completed") return "This circle has completed — every member received the pot once, and clean-completion security deposits were returned.";
  if (f.state === "Defaulted") return "This circle ended in default. Remaining funds and deposits were distributed pro-rata to members who had not yet received; the rotation did not finish normally.";
  if (f.state === "Dissolved") return "This circle was dissolved before it started and every deposit was refunded in full.";
  if (f.isDelinquent) {
    return (
      "You are currently marked delinquent (a missed contribution past grace). " +
      "You must cure (re-deposit) before you can receive your payout."
    );
  }
  if (f.hasReceived) return "You have already received your payout for this circle.";
  if (f.roundsUntilYourTurn === 0) {
    return `It's your turn now — you receive the pot of ${f.intendedPotStr}.`;
  }
  if (f.roundsUntilYourTurn !== null) {
    return `Your payout is in ${f.roundsUntilYourTurn} round(s); you'll receive ${f.intendedPotStr}.`;
  }
  return `Circle state is ${f.state}; the pot is ${f.intendedPotStr}.`;
}

// ---------------------------------------------------------------------------
// Shared LLM layer. Used by BOTH the HTTP route (app/app/api/ask) and the MCP
// `ask` tool, so an agent and a human get the same answer for the same question.
// ---------------------------------------------------------------------------

export type ChatTurn = { role: "user" | "assistant"; content: string };

/** Why an answer came back deterministic — surfaced to the caller instead of failing silently. */
export type AnswerMode = "llm" | "baseline";
export type AnswerResult = { answer: string; mode: AnswerMode; reason?: string };

// History is attacker-controlled input to a paid API, so it is bounded on both axes.
export const MAX_HISTORY_TURNS = 6;
export const MAX_HISTORY_CHARS = 400;

/**
 * Coerce untrusted `history` into at most MAX_HISTORY_TURNS well-formed turns.
 * Anything malformed is dropped rather than rejected — a bad history should degrade the answer,
 * never fail the request.
 */
export function sanitizeHistory(raw: unknown): ChatTurn[] {
  if (!Array.isArray(raw)) return [];
  const out: ChatTurn[] = [];
  for (const t of raw) {
    const role = (t as ChatTurn)?.role;
    const content = (t as ChatTurn)?.content;
    if (role !== "user" && role !== "assistant") continue;
    if (typeof content !== "string") continue;
    const trimmed = content.trim();
    if (!trimmed) continue;
    out.push({ role, content: trimmed.slice(0, MAX_HISTORY_CHARS) });
  }
  // Keep the MOST RECENT turns — the tail is what a follow-up like "and?" refers to.
  return out.slice(-MAX_HISTORY_TURNS);
}

/** The authoritative fact block handed to the model. Every figure comes from chain state. */
export function contextFor(f: MemberFacts, baseline: string): string {
  return `FACTS (authoritative, from chain):
- circle state: ${f.state}
- members joined: ${f.joined} of ${f.slots}
- open slots: ${Math.max(f.slots - f.joined, 0)}
- is member: ${f.isMember}
- has received payout: ${f.hasReceived}
- delinquent: ${f.isDelinquent}
- your payout round (0-indexed): ${f.yourRound ?? "not in the rotation yet"}
- rounds until your turn: ${f.roundsUntilYourTurn ?? "unknown"}
- current recipient: ${f.currentRecipient ?? "none"}
- pot: ${f.intendedPotStr}
- deterministic answer to relay: ${baseline}
`;
}

/**
 * Rephrase the deterministic baseline in the member's language, in context of the conversation.
 *
 * The baseline is passed as authoritative so the model can only explain it, never invent a figure
 * (CLAUDE.md §1.3). With no key — or on any failure — the baseline is returned unchanged, together
 * with a `reason` so a missing key is distinguishable from an expired one.
 */
export async function answerWithLlm(
  question: string,
  facts: MemberFacts,
  opts: { apiKey?: string; history?: ChatTurn[] } = {},
): Promise<AnswerResult> {
  const baseline = baselineAnswer(facts);
  const { apiKey, history = [] } = opts;
  if (!apiKey) return { answer: baseline, mode: "baseline", reason: "no_api_key" };

  try {
    const res = await fetch("https://api.deepseek.com/chat/completions", {
      method: "POST",
      headers: { "content-type": "application/json", authorization: `Bearer ${apiKey}` },
      body: JSON.stringify({
        model: NL_MODEL,
        max_tokens: 200,
        messages: [
          { role: "system", content: SYSTEM_PROMPT },
          ...history,
          { role: "user", content: `${contextFor(facts, baseline)}\nMember asks: ${question}` },
        ],
      }),
    });
    if (!res.ok) {
      const detail = await res.text().catch(() => "");
      console.warn("[nl] llm_http_error", res.status, detail.slice(0, 300));
      return { answer: baseline, mode: "baseline", reason: `llm_http_${res.status}` };
    }
    const data = await res.json();
    const text = (data?.choices?.[0]?.message?.content ?? "").trim();
    if (!text) {
      console.warn("[nl] llm_empty_response");
      return { answer: baseline, mode: "baseline", reason: "llm_empty" };
    }
    return { answer: text, mode: "llm" };
  } catch (err) {
    console.warn("[nl] llm_threw", err);
    return { answer: baseline, mode: "baseline", reason: "llm_threw" };
  }
}

File: agent/src/nl.py
"""Natural-language member Q&A (CLAUDE.md §8) — EN / Nigerian Pidgin / Swahili.

HARD RULE (CLAUDE.md §1.3): the LLM NEVER moves money and never authorizes an action the
contract wouldn't enforce. It only *explains* facts read from chain. Money actions come from
the rule-based `decide()` in loop.py. This module: (1) extracts factual answers from a
CircleView deterministically, (2) optionally phrases them via an LLM in the member's language.
If no LLM key is set, the deterministic answer is returned as-is (still correct, just terser).
"""

from __future__ import annotations

from dataclasses import dataclass

from .chain import CircleView

# NL handler model — small + fast, explanation only (docs/STACK.md). Never tool-enabled.
# DeepSeek (OpenAI-compatible) so the rephrasing runs on the key the deployment holds.
NL_MODEL = "deepseek-chat"
DEEPSEEK_URL = "https://api.deepseek.com/chat/completions"

SYSTEM_PROMPT = """You are AjoAI's member assistant for a rotating savings circle \
(ajo/esusu/chama/stokvel) on Celo. You ONLY explain facts you are given about the circle. \
You NEVER promise to move money, change anyone's turn, waive a penalty, or take any action — \
the smart contract enforces all of that, not you. If asked to do something money-moving, \
explain what the contract rules are instead.

Answer in the SAME language the member used: English, Nigerian Pidgin, or Swahili. Keep it \
short (1-3 sentences), concrete, and money-accurate. Use the exact figures provided. Never \
invent numbers; if a fact is not in the context, say you don't have it."""


def _fmt(amount_wei: int, decimals: int = 18, symbol: str = "units") -> str:
    base = 10**decimals
    whole = amount_wei // base
    frac = amount_wei % base
    s = f"{whole:,}"
    if frac > 0:
        two = (frac * 100) // base  # up to 2 fractional digits
        if two > 0:
            s += "." + f"{two:02d}".rstrip("0")
    return f"{s} {symbol}"


@dataclass
class MemberFacts:
    """Deterministic, chain-derived answer scaffold for one member's common questions."""

    is_member: bool
    has_received: bool
    is_delinquent: bool
    your_round: int | None  # the round index where this member receives
    rounds_until_your_turn: int | None
    current_recipient: str | None
    state: str
    intended_pot_str: str
    joined: int
    slots: int

    def baseline_answer(self) -> str:
        # Non-member: explain the circle's state instead of a flat rejection (the organizer who has
        # not joined yet, an invitee browsing, or someone viewing a finished circle all land here).
        if not is_member_guard(self):
            if self.state == "Forming":
                return (
                    f"This circle is still forming: {self.joined} of {self.slots} have joined. "
                    "You are not in it yet. Tap Join to take a slot (you post a one-round security "
                    "deposit, returned on clean completion)."
                )
            if self.state == "Completed":
                return "This circle has finished. Every member received their payout once and it is now complete."
            if self.state == "Defaulted":
                return "This circle has ended in default. Remaining funds were distributed to members who had not yet received."
            if self.state == "Dissolved":
                return "This circle was dissolved while still forming and all deposits were refunded."
            return f"This circle is active with {self.slots} members; you are not a member of it."
        # Terminal state (member): NEVER project a future round — the rotation is over. `current_round`
        # is stale here, so the round arithmetic below would mis-tell a member "your payout is in N
        # rounds" for a circle that already ended (money-inaccurate, CLAUDE.md §8). State the outcome.
        if self.state == "Completed":
            return "This circle has completed — every member received the pot once, and clean-completion security deposits were returned."
        if self.state == "Defaulted":
            return (
                "This circle ended in default. Remaining funds and deposits were distributed pro-rata "
                "to members who had not yet received; the rotation did not finish normally."
            )
        if self.state == "Dissolved":
            return "This circle was dissolved before it started and every deposit was refunded in full."
        if self.is_delinquent:
            return (
                "You are currently marked delinquent (a missed contribution past grace). "
                "You must cure (re-deposit) before you can receive your payout."
            )
        if self.has_received:
            return "You have already received your payout for this circle."
        if self.rounds_until_your_turn == 0:
            return f"It's your turn now — you receive the pot of {self.intended_pot_str}."
        if self.rounds_until_your_turn is not None:
            return (
                f"Your payout is in {self.rounds_until_your_turn} round(s); "
                f"you'll receive {self.intended_pot_str}."
            )
        return f"Circle state is {self.state}; the pot is {self.intended_pot_str}."


def is_member_guard(f: MemberFacts) -> bool:
    return f.is_member


def facts_for(view: CircleView, member: str, token_decimals: int = 18) -> MemberFacts:
    """Pure: derive a member's situation from a chain snapshot (no LLM, no tx)."""
    member = member.lower()
    members_lc = [m.lower() for m in view.members]
    is_member = member in members_lc

    your_round = None
    if view.rotation:
        rot_lc = [r.lower() for r in view.rotation]
        if member in rot_lc:
            your_round = rot_lc.index(member)

    rounds_until = None
    if your_round is not None:
        rounds_until = max(your_round - view.current_round, 0)

    # has_received: we don't have it on the view per-member; infer for the common case.
    has_received = your_round is not None and view.current_round > your_round

    return MemberFacts(
        is_member=is_member,
        has_received=has_received,
        is_delinquent=(view.recipient is not None and view.recipient.lower() == member and view.recipient_delinquent),
        your_round=your_round,
        rounds_until_your_turn=rounds_until,
        current_recipient=view.recipient,
        state=view.state_name,
        intended_pot_str=_fmt(view.intended_pot, token_decimals),
        joined=len(view.members),
        slots=view.slots,
    )


def answer(question: str, facts: MemberFacts, api_key: str | None) -> str:
    """Phrase the deterministic answer in the member's language via an LLM (optional).

    Uses DeepSeek's OpenAI-compatible chat-completions endpoint. The deterministic baseline is
    passed as authoritative, so the model can only rephrase (never invent a money fact, §1.3).
    Any failure falls back to the baseline, so a missing/invalid key never breaks the answer.
    """
    baseline = facts.baseline_answer()
    if not api_key:
        return baseline

    try:
        import requests

        context = (
            f"FACTS (authoritative, from chain):\n"
            f"- circle state: {facts.state}\n"
            f"- is member: {facts.is_member}\n"
            f"- has received payout: {facts.has_received}\n"
            f"- delinquent: {facts.is_delinquent}\n"
            f"- rounds until your turn: {facts.rounds_until_your_turn}\n"
            f"- pot: {facts.intended_pot_str}\n"
            f"- deterministic answer to relay: {baseline}\n"
        )
        res = requests.post(
            DEEPSEEK_URL,
            headers={"content-type": "application/json", "authorization": f"Bearer {api_key}"},
            json={
                "model": NL_MODEL,
                "max_tokens": 200,
                "messages": [
                    {"role": "system", "content": SYSTEM_PROMPT},
                    {"role": "user", "content": f"{context}\nMember asks: {question}"},
                ],
            },
            timeout=20,
        )
        if not res.ok:
            return baseline
        text = (res.json().get("choices", [{}])[0].get("message", {}).get("content") or "").strip()
        return text or baseline
    except Exception:  # noqa: BLE001 — any LLM/transport error -> safe deterministic baseline
        return baseline
